text/2952-futures.md

```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
-kind of "independent knob" that can be applied orthogonally from other concerns;
-with the two-trait approach, it's "mixed in". And both of the two-trait
-approaches run up against compiler limitations at the moment, though of course
-that shouldn't be taken as a deciding factor.
-
-**The primary reason this RFC opts for the one-trait approach is that it's the
-conservative, forward-compatible option, and has proven itself in practice**.
-It's possible to add `MoveFuture`, together with a blanket impl, at any point in the future.
-Thus, starting with just the single `Future` trait as proposed in this RFC keeps our options
-maximally open while we gain experience.
-
-## Rationale, drawbacks and alternatives to the wakeup design (`Waker`)
-
-Previous iterations of this proposal included a separate wakeup type,
-`LocalWaker`, which was `!Send + !Sync` and could be used to implement
-optimized executor behavior without requiring atomic reference counting
-or atomic wakeups. However, in practice, these same optimizations are
-available through the use of thread-local wakeup queues, carrying IDs
-rather than pointers to wakeup objects, and tracking an executor ID or
-thread ID to perform a runtime assertion that a `Waker` wasn't sent across
-threads. For a simple example, a single thread-locked executor with zero
-atomics can be implemented as follows:
+總而言之，futures 的主要替代模型是 callback-base 的方法，在發現當前方法之前嘗試了 callback-base 幾個月。根據我們的經驗，callback 方法在 Rust 中有幾個缺點：
+
+- 它幾乎在所有地方都被強制分配，因此與 no_std 不相容。
+- 它使取消變得**非常**困難，而對於提議的模型，它只是「刪除」。
+- 主觀上，組合器程式碼非常繁瑣，而基於任務的模型則可以輕鬆且快速的達成。
+
+附帶的 [RFC](https://github.com/rust-lang/rfcs/pull/2394) 中提供了整個 async/await 專案的一些其他先備知識和基本原理。
+
+在本節的其餘部分，我們將深入探討特定的 API 設計問題，其中該 RFC 與 futures 0.2 不同。
+
+## 移除內嵌錯誤的基本原理、缺點和替代方法
+
+在主要特徵中刪除內嵌（build-in）錯誤型別有多種原因：
+
+- **改進的類型檢查和推斷**。錯誤型別是當今使用 futures 組合器時最大的痛點之一，無論是試圖讓不同的型別配對，還是在一段代碼無法產生錯誤時導致推理失敗。需要清楚的是，當 `async` 語法可用時，這些問題將變得不那麼明顯。
+
+- **非同步函式**。如果我們保留一個內嵌的錯誤型別，那麼 `async fn` 應該如何工作就不太清楚了：它是否應該總是要求返回的型別是一個 `Result`？如果不是，當返回非 `Result` 型別時會發生什麼？
+
+- **組合器清晰度**。透過它們是否依賴錯誤來拆分組合器可以闡明語義。尤其對於 stream 來說更是如此，其中錯誤處理是常見的混淆來源。
+
+- **正交性**。一般來說，產生和處理錯誤與核心輪詢機制是分開的，所以在同等條件的情況下，遵循 Rust 的一般設計原則，透過與 `Result` **組合**來處理錯誤似乎很好。
+
+綜上所述，即使使用 `TryFuture`，錯誤嚴重的程式碼也有真正的缺點：
+
+- 需要額外的引入（如果程式碼引入 future prelude，則可以避免，我們也許可以更直接地鼓勵你這麼做）。
+
+- 對於程式來說，受一個特徵的**約束**而**實作**另一個特徵可能會令人困惑。
+
+該 RFC 的錯誤處理部分與其他部分是分開的，因此主要的替代方案是保留內嵌的錯誤型別。
+
+## 核心特徵設計的基本原理、缺點和替代方案 (wrt `Pin`)
+
+撇開上面討論過的正交錯誤處理不談，這個 RFC 中主要的另一個重要項目是核心輪詢方法的 `Pin` 轉移，以及它與 `Unpin`/手動撰寫的 futures 之間的關係。在 RFC 討論過程中，我們基本上確定了解決這個問題的三種主要方法：
+
+- **一個核心特徵**。這就是主要 RFC 文本中採用的方法：只有一個核心 `Future` 特徵，它適用於 `Pin<&mut Self>`。另外還有一個 `poll_unpin` 輔助器，用於在手動實現中使用 `Unpin` futures。
+
+- **兩個核心特徵**。我們可以提供兩個特徵，例如 `MoveFuture` 和 `Future`，其中一個在 `&mut self` 上運行，另一個在 `Pin<&mut Self>` 上運行。這使得我們可以繼續以 futures 0.2 的風格編寫程式，即無需引入 `Pin`/`Unpin` 或以其他方式談論 pin。一個關鍵要求是需要互操作性（interoperation），以便可以在任何需要 `Future` 的地方使用 `MoveFuture`。至少有兩種方法可以實現這種互操作：
```

Comment:
Not sure about **互操作性**. 協同工作性 seems not good either. Do anyone get a reference for it?

Reply:
"相互運作" is colloquialism.
https://terms.naer.edu.tw/search/?q=interoperation&field=ti&op=AND&match=&q=&field=ti&op=AND&order=cat&num=10&show=&page=5&group=1&heading=#result

Reply:
It seems 互操作性 is widely accepted in simplified chinese. I think it's reasonable since we usually using  "互動的" <---> "interactive". I would prefer 可交互運作性 here.
https://terms.naer.edu.tw/detail/3263013/ 可互操作性 interoperation
https://terms.naer.edu.tw/detail/2914181/ 互通性 Interoperability